### ISSUE TITLE
Performance review

### ISSUE THREAD

**Issue by SeanFeldman** [Particular/NServiceBus.AzureStorageQueues#67]
Connects to https://github.com/Particular/PlatformDevelopment/issues/723

The End2End tests show a drop in a performance. An ASQ Transport specialist should take a look into it and profile the transport. The outcome of this could be a list of fixes that need to be applied and/or documentation that needs to updated/created to point users how to improve performance.

Comparison was done between v6 and v7 using EndToEnd testing.
https://particularsoftware.slack.com/archives/tf-v6-perf-tests/p1460293006000468

Consider following hints when working on this:
- is `JsonSerializer` explicitly set for the ASQ transport. Before `core v6`, transports could augment the serialization choice and `ASQ` was changing it to `JSON`. Currently, it needs to be stated explicitly using ` .SerializeMessageWrapperWith<JsonSerializer>();`
- max number of concurrent http connections using either [config section](https://msdn.microsoft.com/en-us/library/fb6y0fyc.aspx) or [System.Net.ServicePointManager.DefaultConnectionLimit](http://msdn.microsoft.com/en-us/library/system.net.servicepointmanager.defaultconnectionlimit.aspx)
- batch size

`/cc @Particular/azure-maintainers

**Comment by Scooletz** [comment]
This is just a remark. We still need the tests.

---
The biggest gain I can think of is `NACK`ing messages using `UpdateMessage` to descrease the `visibilityTimeout` when a failure occurs. In v6 message is invisible for the default invisibility time, currently, we're going to retry much faster.

**Comment by andreasohlund** [comment]
Will this be needed for beta or rc?

**Comment by WilliamBZA** [comment]
@Scooletz @SeanFeldman the code for the tests can be seen over [here](https://github.com/Particular/EndToEnd/pull/8). We'll update this issue once we get more results.

**Comment by Scooletz** [comment]
@WilliamBZA Great!

**Comment by tmasternak** [comment]
@Particular/tf-v6-launch-plat-dev I think that we need to flesh out the scope of this a bit better. That said I think this needs to go into `V6 Launch`

**Comment by boblangley** [comment]
@tmasternak I agree. @andreasohlund asks my question. Perhaps part of finding the answer would be the questions: How bad is the performance drop? Could we go to beta as is?

**Comment by Scooletz** [comment]
@tmasternak @boblangley Then I'd propose just to move to the beta preparation and deal with it.

**Comment by timbussmann** [comment]
since this is part of the the connected issue, does this need a v6 label? I think we can remove this?

quick status on @danielmarbach and my results today:
We could see a huge performance difference between running with the azure storage emulator and a real storage queues. Interestingly, v6 outperforms v5 on the local machine while v5 is clearly "faster" than v6 connecting to real queues.

Some numbers on daniels machine using the GatedSendLocal test from the performance tests:
Storage Emulator:
* v5: Avg: ~41 msg/sec, (send) batchsize 256
* v6: Avg: ~104 msg/sec, (send) batchsize 1024

Azure Storage Queues:
* v5: Avg: ~19 msg/sec, (send) batchsize: 64
* v6: Avg: ~8 msg/sec, (send) batchsize: 512

The difference in batchsize shows, that v6 in both scenario sends messages much faster than v5 but seems to have difficulties with consuming them.

**Comment by SeanFeldman** [comment]
Could someone let know why we're testing against the emulator? Emulator is never used in production, it's a subset of the real service (not the same exceptions and internal complexity), and it's **local**, which is never the case fo Azurer storage service.

**Comment by danielmarbach** [comment]
We are not testing against the emulator. We used it for a few runs to
understand the current solution until the necessary storage and ressource
groups were deployed

**Comment by danielmarbach** [comment]
Did some thorough analysis over the weekend.

* Tuning ServicePointManager is crucial for V5 and V6

```
             ServicePointManager.DefaultConnectionLimit = 5000; // default settings only allows 2 concurrent requests per process to the same host
            ServicePointManager.UseNagleAlgorithm = false; // optimize for small requests
            ServicePointManager.Expect100Continue = false; // reduces number of http calls
            ServicePointManager.CheckCertificateRevocationList = false; // optional, only disable if all dependencies are trusted 
```

* Batch size is crucial, default in v5 is 10. Default in v6 is 10 as well. I think we should put it to 32 by default
* Azure SDK async performance did not degrade between 4.3.0 (used in V5) and 7.0.0 (used in V6). I tested it by downgrading v6 to 4.3.0
* Sync API is not performing better than async API
* V6 deserialization magic and string replacing is more costly than v5 version in terms of allocation but this doesn't make the key difference
* Did several optimizations including late deserialization of message envelope including deserialization (CPU bound) in parallel which didn't bring the dramatic speed improvements
* The crucial difference is how v5 treats concurrency. It treats concurrency as parallelism and gets messages from the queue in parallel. The v6 version uses a single message pump thread to get messages from the queue and then dispatches the message up to the defined concurrency limit concurrently. Tweaking the batch size and the concurrency limit doesn't help much. The limiting factor is the queueClient.GetMessages or GetMessagesAsync combined with a sensible default in concurrency. Since ASQ in contrast to ASB doesn't offer a push based API we need to query in parallel to reach max performance. 

Using my own storage account in North Europe. 

```
V5 (latest with latest 4.3.0 ASQ)
Concurrency here means threads

1000 msg / concurrency 32 / batch 32
Sending #1000 of msgs over the bus took 0:00:00:08.3455026
Receiving #1000 of msgs over the bus took 0:00:00:05.6959472

1000 msg / concurrency 100 / batch 32
Receiving #1000 of msgs over the bus took 0:00:00:02.6269757

1000 msg / concurrency 200 / batch 32
Receiving #1000 of msgs over the bus took 0:00:00:04.1939380

V6 (develop)

Concurrency here means number of messages

1000 msg / concurrency 32 /batch 32
Sending #1000 of msgs over the bus took 0:00:00:10.6905027
Receiving #1000 of msgs over the bus took 0:00:00:07.9081931

1000 msgs/ concurrency 100 / batch 32
Receiving #1000 of msgs over the bus took 0:00:00:07.1566520

1000 msgs/ concurrency 200 / batch 32
Receiving #1000 of msgs over the bus took 0:00:00:07.4117855

V6 (MessagePump)

1000 msg / concurrency 32 /batch 32
Receiving #1000 of msgs over the bus took 0:00:00:06.6142550

1000 msgs/ concurrency 100 / batch 32
Receiving #1000 of msgs over the bus took 0:00:00:07.6507209

1000 msgs/ concurrency 200 / batch 32
Receiving #1000 of msgs over the bus took 0:00:00:07.1288474

V6 (MessagePump in Parallel 8)

1000 msgs / concurrency 32 / batch 32
Receiving #1000 of msgs over the bus took 0:00:00:02.6774764

1000 msgs/ concurrency 100 / batch 32
Receiving #1000 of msgs over the bus took 0:00:00:01.8962020

1000 msgs / concurrency 200 / batch 32
Receiving #1000 of msgs over the bus took 0:00:00:02.4874563

V6 (MessagePump in Parallel 16)

1000 msgs / concurrency 32 / batch 32
Receiving #1000 of msgs over the bus took 0:00:00:02.6054616

1000 msgs/ concurrency 100 / batch 32
Receiving #1000 of msgs over the bus took 0:00:00:01.9328415

1000 msgs / concurrency 200 / batch 32
Receiving #1000 of msgs over the bus took 0:00:00:02.0487322

V6 (MessagePump in Parallel, Unpack Parallel)

1000 msgs / concurrency 32 / batch 32
Receiving #1000 of msgs over the bus took 0:00:00:02.4768056

1000 msgs/ concurrency 100 / batch 32
Receiving #1000 of msgs over the bus took 0:00:00:01.8923439

1000 msgs / concurrency 200 / batch 32
Receiving #1000 of msgs over the bus took 0:00:00:02.0572539
```

Sending in spike PRs soon

**Comment by danielmarbach** [comment]
The receive performance [PR sent in.](https://github.com/Particular/NServiceBus.AzureStorageQueues/pull/79) Needs review and merge

**Comment by timbussmann** [comment]
publisher performance PR here: https://github.com/Particular/NServiceBus.AzureStorageQueues/pull/91

**Comment by danielmarbach** [comment]
With the new parallelism approach we have now the following "dilemma":

With previous versions of NSB.ASQ and NServiceBus v5 the default concurrency limit was set to 1 (one). Now with our algorithm the default parallelism will be 10 (ten). So that means when the queue is idle still 10 processing loops will try to receive messages and then idle for the defined amount of time. So that means in the default configuration we increased the running costs by ten fold. We did a quick math on the latest pricings for storage. Instead of costing ~ 8 cents per endpoint per month the costs are now 80 cents per endpoint per month. Since the default concurrency setting in previous versions of the transport were not useful in practice anyway the costs there also increase by the concurrency settings.

We decided to not implement any backtrack algorithm and document that fact on the ASQ performance tuning documentation site. We can easily implement a back tracking algorithm internally in a minor version if necessary. Discussed with @timbussmann and @yvesgoeleven 

// cc @Particular/azure-storage-queues-maintainers
